src/validation.cpp

```diff
@@ -3681,6 +3694,51 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
     return true;
 }
 
+bool ReplayBlocks(const CChainParams& params, CCoinsView* view)
+{
+    LOCK(cs_main);
+
+    CCoinsViewCache cache(view);
+
+    uint256 hashBest = view->GetBestBlock();
+    uint256 hashUpto = view->GetUptoBlock();
+
+    if (hashUpto.IsNull() || hashBest == hashUpto) return true;
+
+    if (mapBlockIndex.count(hashUpto) == 0) {
+        return error("ReplayBlocks(): chainstate boundaries not in block index");
+    }
+    auto pindexUpto = mapBlockIndex[hashUpto];
+
+    int nHeight = 1; // Skip the genesis block
+    if (mapBlockIndex.count(hashBest)) {
+        auto pindexBest = mapBlockIndex[hashBest];
+        if (pindexUpto->GetAncestor(pindexBest->nHeight) != pindexBest) {
+            return error("ReplayBlocks(): chainstate tip does not derive from final boundary");
+        }
+        nHeight = std::max(nHeight, pindexBest->nHeight);
+    }
+
+    uiInterface.ShowProgress(_("Replaying blocks..."), 0);
+    LogPrintf("Replaying %i blocks\n", pindexUpto->nHeight - nHeight);
+
+    while (nHeight < pindexUpto->nHeight) {
+        ++nHeight;
+        auto pindex = pindexUpto->GetAncestor(nHeight);
+        CBlock block;
+        if (!ReadBlockFromDisk(block, pindex, params.GetConsensus())) {
+            return error("ReplayBlocks(): ReadBlockFromDisk failed at %d, hash=%s", pindex->nHeight, pindex->GetBlockHash().ToString());
+        }
+        ReplayBlock(block, cache, pindex->nHeight);
+    }
+    cache.SetBestBlock(hashUpto);
+    chainActive.SetTip(pindexUpto);
```

Comment:
It seems super weird to be acting entirely on non-globals, and then suddenly set a global here.